### PR TITLE
Use prefix extractor to improve performance

### PR DESCRIPTION
**Pull request by git-hulk**
- [x] Add benchmark result to confirm that can improve the performance

Update:

The RocksDB prefix extractor looks like that can bring significant performance improvement, but NOT for Kvrocks from my side. Welcome to have a try and correct me if I was wrong.

### References

- [RocksDB Prefix Seek API](https://github.com/EighteenZi/rocksdb_wiki/blob/master/Prefix-Seek-API-Changes.md)
- [Adopting RocksDB within Manhattan](https://blog.twitter.com/engineering/en_us/topics/infrastructure/2021/adopting-rocksdb-within-manhattan)
- [RocksDB in TiKV](https://tikv.org/blog/rocksdb-in-tikv/)
- [How We Use RocksDB at Rockset](https://rockset.com/blog/how-we-use-rocksdb-at-rockset/)
* [Tuning RocksDB - Prefix Extractor](https://www.jianshu.com/p/26214e45fd4a)

### CODE REVIEW

File: src/storage/prefix_extractor.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    ptr += 4 + key_size;  // drop key encoding size and key
+    // return the origin key if it's too short
+    if (ptr > key.data() + key.size()) return key;
+
+    size_t prefix_len = 1 /* ns encoding size */ + ns_size + 4 /* encoding key size */ + key_size;
+    if (cluster_enabled_) {
+      prefix_len += 2;  // add slot id encoding size if the cluster mode is enabled
+    }
+    return {key.data(), prefix_len};
```
Comment (PragmaTwice) [inline src/storage/prefix_extractor.h]:
```suggestion
    return {key.data(), ptr - key.data()};
```

**Comment by git-hulk** [comment]
We didn't add the version into the prefix, since it may cause bugs in some test cases like [hash_test.cc#L203](https://github.com/apache/incubator-kvrocks/blob/unstable/tests/cppunit/types/hash_test.cc#L203):
```C++
/Users/hulk/code/cxx/incubator-kvrocks/tests/cppunit/types/hash_test.cc:315: Failure
Expected equality of these values:
  4 + 26
    Which is: 30
  result.size()
    Which is: 0
```

And found it was caused by the upper bound in `SeekToLast` after diving into the reason. The root cause is the `Seek` API will also use the bloom filter to probe the existence of the prefix when enabling the prefix extractor, then it will be false if we added the version into the prefix when seeking with the upper bound(which is the version + 1).

```C++
void DBIter::SeekToLast() {
  if (iterate_upper_bound_ != nullptr) {
    // Seek to last key strictly less than ReadOptions.iterate_upper_bound.
    SeekForPrev(*iterate_upper_bound_);
    const bool is_ikey = (timestamp_size_ > 0 && timestamp_lb_ != nullptr);
    Slice k = Valid() ? key() : Slice();
    if (is_ikey && Valid()) {
      k.remove_suffix(kNumInternalBytes + timestamp_size_);
    }
```

We can work around this issue by removing the version like the current implementation, but adding the version would help a lot if there have many deleted tombstones. So another way is to avoid using the upper bound when using `SeekToLast`.

**Comment by git-hulk** [comment]
Update: I have tested if the prefix extractor can bring better performance when seeking, but the result doesn't have any difference while comparing to adding the version into the prefix. Also, avoiding using the upper bound would introduce some complexity in the codebase, so let's remove the version from the prefix extractor.

The performance benefit of the prefix extractor, I tested with below random keys on my MacBook, also can't see the improvement:

```
# Create Hash keys with hset 

❯ redis-benchmark -p 6666 -n 1000000 -r 50000 hset key__rand_int__ field__rand_int__ value

# Benchmark with HRANGEBYLEX

❯ redis-benchmark -p 6666 -n 1000000 -r 50000 HRANGEBYLEX key__rand_int__ "[field000000000000" "[field000000005000"

```

So we can keep this PR in draft mode, and users can have a test if they want to try.

**Comment by mapleFU** [comment]
Do we have some benchmark or micro-benchmark here? Seems MyRocks requires that, so RocksDB implements it. Can we take some benchmark from MyRocks?

**Comment by git-hulk** [comment]
Thanks @mapleFU, do you have any references?

**Comment by mapleFU** [comment]
1. MyRocks Paper, Section 3.2.2.1: https://vldb.org/pvldb/vol13/p3217-matsunobu.pdf
2. https://www.slideshare.net/matsunobu/myrocks-deep-dive/58
3. https://www.cidrdb.org/cidr2017/papers/p82-dong-cidr17.pdf

> Preﬁx Bloom ﬁlters. Bloom ﬁlters do not help with range queries. We have developed a preﬁx Bloom ﬁlter that helps with range queries, based on the observation that many range queries are often over a preﬁx; e.g., the userid part of a (userid,timestamp) key or postid of a (postid,likerid) key. We allow users to deﬁne preﬁx extractors to deterministically extract a preﬁx part of the key from which we construct a Bloom ﬁlter. When querying a range, the user can specify that the query is on a deﬁned preﬁx. We have found this optimization reduces read ampliﬁcation (and attendant CPU overheads) on range queries by up to 64% on our systems.

**Comment by git-hulk** [comment]
Close this PR since we cannot find any performance improvements. Can reopen it if anyone is interested in this.

**Pull request closed without merging**